### DISCUSSION TITLE
OSPP 2023 Challenge - for #2438

### DISCUSSION THREAD

**Discussion by q82419** [WasmEdge/WasmEdge#2453] [Show and tell]
## Introduction

Dear Applicants,

We are excited to announce the pre-test for the OSPP project #2438, "Expanding WasmEdge WASI Socket Support to macOS". This project aims to evaluate your skills and qualifications for the program, and we encourage all interested candidates to participate.

## Requirements

1. Please complete the following tasks on a Apple MacOS platforms (no matter Intel or Apple Silicon):
  A. Use the Apple Clang compiler to build WasmEdge with the tests. Please refer to [our workflow files](https://github.com/WasmEdge/WasmEdge/blob/master/.github/workflows/reusable-build-on-macos.yml).
  B. Add a new WASI host function (you can determine the export name) for return a value with adding 2 input values.
  C. BONUS: It'll be better if you add the unit tests in the [test file](https://github.com/WasmEdge/WasmEdge/tree/master/test/host/socket).

2. Once you have completed the tasks, please provide a link to your GitHub repository containing the following:
  A. A README file explaining how you approached the tasks and documenting your results.
  B. Screenshots or logs of the results for running the tests on MacOS.
  C. BONUS: Screenshots or logs of running the tests in 1C. 

## Deadline

Please refer to [the OSPP 2023 timeline](https://summer-ospp.ac.cn/?lang=en) for the final deadline for valid submissions, which is June 4th, UTC+8.

We look forward to reviewing your submissions and thank you for your interest.

**Comment by MediosZ** [comment]
Hi, I tried fixing the socket tests on macOS, here's the link to [the branch](https://github.com/MediosZ/WasmEdge/tree/macos-networking).

`sizeof(sockaddr_in) - sizeof(sockaddr_in::sin_family);` on Linux is 14 but on macOS it's 15.
That's because  `sin_family` on Linux is `unsigned long`, but `__uint8_t` on macOS.
So I add a condition to handle this on both platforms.
Now the socket tests pass on macOS.
Here's the log:
```
➜  build git:(macos-networking) ./test/host/socket/wasiSocketTests 
[==========] Running 9 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 9 tests from WasiSockTest
[ RUN      ] WasiSockTest.SocketUDP_4V1
[       OK ] WasiSockTest.SocketUDP_4V1 (0 ms)
[ RUN      ] WasiSockTest.SocketUDP_4V2
[       OK ] WasiSockTest.SocketUDP_4V2 (0 ms)
[ RUN      ] WasiSockTest.SocketUDP_6
[       OK ] WasiSockTest.SocketUDP_6 (0 ms)
[ RUN      ] WasiSockTest.SocketUDP_4_Fallback
[       OK ] WasiSockTest.SocketUDP_4_Fallback (0 ms)
[ RUN      ] WasiSockTest.SocketUDP_6_Fallback
[       OK ] WasiSockTest.SocketUDP_6_Fallback (0 ms)
[ RUN      ] WasiSockTest.SockOpt
[       OK ] WasiSockTest.SockOpt (0 ms)
[ RUN      ] WasiSockTest.SockGetLocalAddr_4
[       OK ] WasiSockTest.SockGetLocalAddr_4 (0 ms)
[ RUN      ] WasiSockTest.SockGetLocalAddr_6
[       OK ] WasiSockTest.SockGetLocalAddr_6 (0 ms)
[ RUN      ] WasiSockTest.GetAddrinfo
[       OK ] WasiSockTest.GetAddrinfo (54 ms)
[----------] 9 tests from WasiSockTest (57 ms total)

[----------] Global test environment tear-down
[==========] 9 tests from 1 test suite ran. (57 ms total)
[  PASSED  ] 9 tests.
```

I am using `Apple clang version 14.0.0 (clang-1400.0.29.202)` on macOS 12.6.3
